src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs

```diff
@@ -530,5 +535,95 @@ public static string[] GetLogicalDrives()
         {
             return DriveInfoInternal.GetLogicalDrives();
         }
+
+        internal static string? GetLinkTarget(ReadOnlySpan<char> linkPath, bool isDirectory) => Interop.Sys.ReadLink(linkPath);
+
+        internal static void CreateSymbolicLink(string path, string pathToTarget, bool isDirectory)
+        {
+            string pathToTargetFullPath = PathInternal.GetLinkTargetFullPath(path, pathToTarget);
+
+            // Fail if the target exists but is not consistent with the expected filesystem entry type
+            if (Interop.Sys.Stat(pathToTargetFullPath, out Interop.Sys.FileStatus targetInfo) == 0)
+            {
+                if (isDirectory != ((targetInfo.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFDIR))
+                {
+                    throw new IOException(SR.Format(SR.IO_InconsistentLinkType, path));
+                }
+            }
+
+            Interop.CheckIo(Interop.Sys.SymLink(pathToTarget, path), path, isDirectory);
+        }
+
+        internal static FileSystemInfo? ResolveLinkTarget(string linkPath, bool returnFinalTarget, bool isDirectory)
+        {
+            // throws if the current link file does not exist
+            Interop.CheckIo(Interop.Sys.LStat(linkPath, out Interop.Sys.FileStatus info), linkPath, isDirectory);
+
+            if ((info.Mode & Interop.Sys.FileTypes.S_IFMT) != Interop.Sys.FileTypes.S_IFLNK)
+            {
+                return null;
+            }
+
+            ValueStringBuilder sb = new(stackalloc char[Interop.DefaultPathBufferSize]);
+            sb.Append(linkPath);
+
+            if (returnFinalTarget)
+            {
+                int visitCount = 0;
+                while (true)
+                {
+                    if (!TryGetLinkTarget(ref sb))
+                    {
+                        // We finally found the final target: either
+                        // this file does not exist (broken links are acceptable)
+                        // or this file is not a link
+                        break;
+                    }
+
+                    visitCount++;
+                    if (visitCount > MaxFollowedLinks)
+                    {
+                        // We went over the limit and couldn't reach the final target
+                        throw new IOException(SR.Format(SR.IO_TooManySymbolicLinkLevels, linkPath));
+                    }
+                }
+            }
+            else
+            {
+                TryGetLinkTarget(ref sb);
```

Comment:
Is it possible for `TryGetLinkTarget` to return `false` here? If so, the code below is going to return `new DirectoryInfo(linkPath)` or `new FileInfo(linkPath)` instead of `null`

Reply:
Yes.

Reply:
Intention is to return source path if no target exists.

Reply:
Between the `lstat` file and `readlink` the file may have changed.
We can omit the `lstat` call and just call `readlink`. When the target is not a link we can detect that as `errno == EINVAL`.
Besides getting rid of the race, it saves us a syscall too.

Reply:
Done, we no longer call lstat after https://github.com/dotnet/runtime/pull/54253/commits/3e36f72df85605aa37e1e8c8b07729f56d05eb77. However when returnFinalTarget is true, we still will do multiple calls to readlink in order to get the final target. The multiple syscalls in this case can probably be reduced by using realpath but @carlossanlop told me that he found some issues while trying to implement that. I can give it a try to realpath in a follow-up PR if that's OK with you.

Reply:
> e still will do multiple calls to readlink in order to get the final target. The multiple syscalls in this case can probably be reduced by using realpath

`realpath` is not a syscall (on Linux), it implemented in the c library and calls `readlink`. I don't expect it to perform better.

Reply:
In fact, `realpath` is more expensive than what you are doing here, because it will resolve any links in the parent directories also.